dspace-api/src/main/java/org/dspace/app/bulkedit/MetadataExport.java

```diff
@@ -7,272 +7,101 @@
-        // print the help message
-        HelpFormatter myhelp = new HelpFormatter();
-        myhelp.printHelp("MetadataExport\n", options);
-        System.out.println("\nfull export: metadataexport -f filename");
-        System.out.println("partial export: metadataexport -i handle -f filename");
-        System.exit(exitCode);
-    }
-
-    /**
-     * main method to run the metadata exporter
-     *
-     * @param argv the command line arguments given
-     * @throws Exception if error occurs
-     */
-    public static void main(String[] argv) throws Exception {
-        // Create an options object and populate it
-        CommandLineParser parser = new PosixParser();
-
+    private Options constructOptions() {
         Options options = new Options();
 
         options.addOption("i", "id", true, "ID or handle of thing to export (item, collection, or community)");
+        options.getOption("i").setType(String.class);
         options.addOption("f", "file", true, "destination where you want file written");
+        options.getOption("f").setType(OutputStream.class);
+        options.getOption("f").setRequired(true);
         options.addOption("a", "all", false,
                           "include all metadata fields that are not normally changed (e.g. provenance)");
+        options.getOption("a").setType(boolean.class);
         options.addOption("h", "help", false, "help");
+        options.getOption("h").setType(boolean.class);
 
-        CommandLine line = null;
-
-        try {
-            line = parser.parse(options, argv);
-        } catch (ParseException pe) {
-            System.err.println("Error with commands.");
-            printHelp(options, 1);
-            System.exit(0);
-        }
 
-        if (line.hasOption('h')) {
-            printHelp(options, 0);
-        }
+        return options;
+    }
 
-        // Check a filename is given
-        if (!line.hasOption('f')) {
-            System.err.println("Required parameter -f missing!");
-            printHelp(options, 1);
+    public void internalRun() throws Exception {
+        if (help) {
+            handler.logInfo("\nfull export: metadata-export -f filename");
+            handler.logInfo("partial export: metadata-export -i handle -f filename");
+            printHelp();
+            return;
         }
-        String filename = line.getOptionValue('f');
 
-        // Create a context
-        Context c = new Context(Context.Mode.READ_ONLY);
-        c.turnOffAuthorisationSystem();
-
-        // The things we'll export
-        Iterator<Item> toExport = null;
-        MetadataExport exporter = null;
+        DSpaceCSV dSpaceCSV = metadataDSpaceCsvExportService
+            .handleExport(context, exportAllItems, exportAllMetadata, handle,
+                          handler);
+        handler.writeFilestream(context, filename, dSpaceCSV.getInputStream(), "exportCSV");
+        context.restoreAuthSystemState();
+        context.complete();
```

Comment:
Yes, this introduces a subtle coupling between the creating class and the completing class that could bit us later.

Reply:
Again, please resolve prior feedback here.

Reply:
@tdonohue @mwoodiupui This is not something that we can easily dismiss here. This class represents a class that is run in a separate thread. This class creates the context & closes it as well. For this particular case we could go for a context.abort() as well (as this is an export that doesn't modify any data). But leaving it out would not be recommended as we don't want any open database connections for threads that are done.

Reply:
@KevinVdV : I'm sorry, but we cannot merge this as is. It will need refactoring.  Currently you've changed the behavior of this `MetadataExport` class, which is fine.  But, you've added a major flaw here.  Essentially, there are now two public methods... a `setup()` which creates a Context, and a `internalRun()` which commits the Context.  If someone accidentally calls `internalRun()` **first**, this will immediately fail with a NPE as no Context yet exists.   My recommendation here is to not have so many public methods, and ensure there is _one method_ that creates & completes the context. It can call other methods to do processing with that context, but the context creation & closing must occur in the same method (no exceptions).

Reply:
Should not `internalRun()` create and adjust the Context, use it, and close it?  Should not `setup()` simply collect information that `internalRun()` will use later?

I must agree with @tdonohue here:  a Context should be created and closed within a single method, preferably within a single `try` block.